bench/macro/lsm-tree-bench-wp8.hs

```diff
@@ -0,0 +1,404 @@
+
+-------------------------------------------------------------------------------
+-- dry-run
+-------------------------------------------------------------------------------
+
+doDryRun :: GlobalOpts -> RunOpts -> IO ()
+doDryRun gopts opts = do
+    time <- timed_ $ doDryRun' gopts opts
+    printf "Batch generation: %.03f sec\n" time
+
+doDryRun' :: GlobalOpts -> RunOpts -> IO ()
+doDryRun' gopts opts = do
+    -- calculated some expected statistics for generated batches
+    id $ do
+        -- we generate n random numbers in range of [ 1 .. d ]
+        -- what is the chance they are all distinct
+        let n = fromIntegral (opts.batchCount * opts.batchSize) :: Double
+        let d = fromIntegral gopts.initialSize :: Double
+        -- this is birthday problem.
+        let p = 1 - exp (negate $  (n * (n - 1)) / (2 * d))
+
+        -- number of people with a shared birthday
+        -- https://en.wikipedia.org/wiki/Birthday_problem#Number_of_people_with_a_shared_birthday
+        let q = n * (1 - ((d - 1) / d) ** (n - 1))
+
+        printf "Probability of a duplicate: %5f\n" p
+        printf "Expected number of duplicates: %5f out of %f\n" q n
+
+    -- TODO: open session to measure that as well.
+    let initGen = SM.mkSMGen opts.seed
+
+    keysRef <- newIORef $
+        if opts.check
+        then IS.fromList [ 0 .. gopts.initialSize - 1 ]
+        else IS.empty
+    duplicateRef <- newIORef (0 :: Int)
+
+    void $ forFoldM_ initGen [ 0 .. opts.batchCount - 1 ] $ \b g -> do
+        let lookups :: [Word64]
+            inserts :: [Word64]
+            (!nextG, lookups, inserts) = generateBatch gopts.initialSize opts.batchSize g b
+
+        when opts.check $ do
+            keys <- readIORef keysRef
+            let new  = IS.fromList $ map fromIntegral lookups
+            let diff = IS.difference new keys
+            -- when (IS.notNull diff) $ printf "missing in batch %d %s\n" b (show diff)
+            modifyIORef' duplicateRef $ \n -> n + IS.size diff
+            writeIORef keysRef $! IS.union
+                (IS.difference keys new)
+                (IS.fromList $ map fromIntegral inserts)
+
+        let (batch1, batch2) = toOperations lookups inserts
+        _ <- evaluate $ force (batch1, batch2)
+
+        return nextG
+
+    when opts.check $ do
+        duplicates <- readIORef duplicateRef
+        printf "True duplicates: %d\n" duplicates
+
+-------------------------------------------------------------------------------
+-- Batch generation
+-------------------------------------------------------------------------------
+
+{- | Implement generation of unbounded sequence of insert/delete operations
+
+matching UTxO style from spec: interleaved batches insert and lookup
+configurable batch sizes
+1 insert, 1 delete, 1 lookup per key.
+
+Current approach is probabilistic, but uses very little state.
+We could also make it exact, but then we'll need to carry some state around
+(at least the difference).
```

Comment:
> rejection should allow us to generate sequences without duplicates.

I'm not sure. I have an better idea, but I don't want stall this PR indefinitely discussing options.

Reply:
Sounds good. To clarify: I don't think a non-probabilistic implementation is necessary to merge this PR, but it would be good to find out soon whether we will need it or not.

Then let's leave the implementation as is and I will make sure the the questions gets answered. If necessary, we can then do it as a follow-up.